File: solution/ts/week03-cqs/exercise2/src/cache.ts
export class Cache {
    private map: Map<string, number> = new Map<string, number>();

    get(key: string): number | undefined {
        return this.map.get(key); // Just returns the value or undefined if not present
    }

    insertIfAbsent(key: string, value: number): void {
        if (!this.map.has(key)) {
            this.map.set(key, value); // Only modifies the state, does not return value
        }
    }

    ensurePresentAndGet(key: string, defaultValue: number): number {
        this.insertIfAbsent(key, defaultValue);
        return this.get(key) ?? defaultValue;
    }
}

File: solution/c#/week03-cqs/Cache/Cache.cs
namespace Cache
{
    public class Cache
    {
        private readonly Dictionary<string, int> _map = new();

        public int? Get(string key)
        {
            _map.TryGetValue(key, out var value);
            return value;
        }

        public void InsertIfAbsent(string key, int value)
        {
            if (!_map.ContainsKey(key))
            {
                _map[key] = value;
            }
        }

        public int EnsurePresentAndGet(string key, int defaultValue)
        {
            InsertIfAbsent(key, defaultValue);
            return Get(key) ?? defaultValue;
        }
    }
}